libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/generator/OpenHouseDataLayoutGenerator.java

```diff
@@ -0,0 +1,113 @@
+  private static final long REWRITE_BYTES_PER_SECOND = 2 * MB;
+  private static final long EXECUTOR_MEMORY_MB = 2048;
+  private static final int MAX_CONCURRENT_FILE_GROUP_REWRITES = 50;
+  private static final int REWRITE_PARALLELISM = 900; // number of Spark tasks to run in parallel
+  private static final long TARGET_BYTES_SIZE = 2 * FILE_BLOCK_SIZE_BYTES - FILE_BLOCK_MARGIN_BYTES;
+  private static final double COMPUTE_STARTUP_COST_GB_HR = 0.5;
+  private final TableFileStats tableFileStats;
+  private final TablePartitionStats tablePartitionStats;
+
+  /**
+   * Generate a list of data layout optimization strategies based on the table file stats and
+   * historic query patterns.
+   */
+  @Override
+  public List<DataLayoutOptimizationStrategy> generate() {
+    // skip single partition and non-partitioned tables
+    if (tablePartitionStats.get().count() <= 1) {
+      return Collections.emptyList();
+    }
+    return Collections.singletonList(generateCompactionStrategy());
+  }
+
+  private DataLayoutOptimizationStrategy generateCompactionStrategy() {
+    List<Long> fileSizes =
+        tableFileStats
+            .get()
+            .map((MapFunction<FileStat, Long>) FileStat::getSize, Encoders.LONG())
+            .filter(
+                (FilterFunction<Long>)
+                    size ->
+                        size < TARGET_BYTES_SIZE * DataCompactionConfig.MIN_BYTE_SIZE_RATIO_DEFAULT)
+            .collectAsList();
+
+    DataCompactionConfig.DataCompactionConfigBuilder configBuilder = DataCompactionConfig.builder();
+
+    configBuilder.targetByteSize(TARGET_BYTES_SIZE);
+
+    long totalSizeBytes = fileSizes.stream().reduce(0L, Long::sum);
+    long estimatedFileGroupsCount =
+        Math.max(
+            tablePartitionStats.get().count(),
+            totalSizeBytes / DataCompactionConfig.MAX_FILE_GROUP_SIZE_BYTES_DEFAULT);
+
+    int maxCommitsCount =
+        (int)
+            Math.min(
+                MAX_NUM_COMMITS,
+                (estimatedFileGroupsCount + NUM_FILE_GROUPS_PER_COMMIT - 1)
+                    / NUM_FILE_GROUPS_PER_COMMIT);
+    configBuilder.partialProgressMaxCommits(maxCommitsCount);
+
+    int estimatedTasksPerFileGroupCount =
+        (int) (DataCompactionConfig.MAX_FILE_GROUP_SIZE_BYTES_DEFAULT / TARGET_BYTES_SIZE);
+
+    configBuilder.maxConcurrentFileGroupRewrites(
+        Math.min(
+            MAX_CONCURRENT_FILE_GROUP_REWRITES,
+            (estimatedTasksPerFileGroupCount + REWRITE_PARALLELISM - 1)
+                / estimatedTasksPerFileGroupCount));
+
+    // don't split large files
+    configBuilder.maxByteSizeRatio(MAX_BYTES_SIZE_RATIO);
+
+    long filesReducedCount = estimateReducedFilesCount(fileSizes);
+    double computeGbHr = estimateComputeGbHr(fileSizes);
+    double filesReducedCountPerComputeGbHr = filesReducedCount / computeGbHr;
+    return DataLayoutOptimizationStrategy.builder()
+        .config(configBuilder.build())
```

Comment:
```suggestion
        .type(PARTITIONED_TABLE_FILE_RIGHT_SIZING_STRATEGY)
        .config(configBuilder.build())
```

Let's introduce a strategy enum for example. 
```
enum {
     PARTITIONED_TABLE_FILE_RIGHT_SIZING_STRATEGY,
     PARTITIONED_TABLE_SORTING_STRATEGY //  (add in future)
}
```

Reply:
It is making it into a persistent layer of tblproperty. Changing is later makes it a bit difficult.

Reply:
Thanks, Sumedh. When strategy list is updated with overwrite, no need to make older version compatible, the future app that reads the state can be disabled during rollout. I'm not sure if it should be type attribute or smth else, let's keep it open for now.